cmd/pitr/collector/collector.go

```diff
@@ -0,0 +1,262 @@
+		pxcPass:        c.PXCPass,
+		pxcServiceName: c.PXCServiceName,
+		bucketPrefix:   prefix,
+	}, nil
+}
+
+func (c *Collector) Run() error {
+	err := c.newDB()
+	if err != nil {
+		return errors.Wrap(err, "new db connection")
+	}
+	defer c.closeDB()
+
+	err = c.CollectBinLogs()
+	if err != nil {
+		return errors.Wrap(err, "collect binlog files")
+	}
+
+	return nil
+}
+
+func (c *Collector) newDB() error {
+	host, err := network.GetPXCLastHost(c.pxcServiceName)
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+	pxc, err := db.NewPXC(host, c.pxcUser, c.pxcPass)
+	if err != nil {
+		return errors.Wrapf(err, "new manager with host %s", host)
+
+	}
+	c.db = pxc
+
+	return nil
+}
+
+func (c *Collector) closeDB() error {
+	return c.db.Close()
+}
+
+func (c *Collector) CollectBinLogs() error {
+	list, err := c.db.GetBinLogList()
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+	// get last uploaded binlog file name
+	binlogName, err := c.db.GetBinLogName(c.lastSet)
+	if err != nil {
+		return errors.Wrap(err, "get binlog name by set")
+	}
+
+	upload := false
+	// if there are no uploaded files we going to upload every binlog file
+	if len(binlogName) == 0 {
+		upload = true
+	}
+
+	for _, binlog := range list {
+		// this check is for uploading starting from needed file
+		c.db.GetGTIDSet(binlog)
+		if binlog == binlogName {
+			bSet, err := c.db.GetGTIDSet(binlog)
+			if err != nil {
+				return errors.Wrap(err, "get binlog gtid set")
+			}
+			if c.lastSet != bSet {
+				upload = true
+			}
+		}
+		if upload {
+			err = c.manageBinlog(binlog)
+			if err != nil {
+				return errors.Wrap(err, "manage binlog")
+			}
+		}
+	}
+
+	return nil
+}
+
+type reader struct {
+	r io.Reader
+}
+
+func (r *reader) Read(p []byte) (int, error) {
+	return r.r.Read(p)
+}
+
+func (c *Collector) manageBinlog(binlog string) (err error) {
+	set, err := c.db.GetGTIDSet(binlog)
+	if err != nil {
+		return errors.Wrap(err, "get GTID set")
+	}
+	if len(set) == 0 {
+		return nil
+	}
+
+	binlogTmstmp, err := c.db.GetBinLogFirstTimestamp(binlog)
+	if err != nil {
+		return errors.Wrapf(err, "get first timestamp for %s", binlog)
+	}
+
+	binlogName := "binlog_" + binlogTmstmp + "_" + fmt.Sprintf("%x", md5.Sum([]byte(set)))
+	if len(c.bucketPrefix) > 0 {
+		binlogName = c.bucketPrefix + "/" + binlogName
+	}
+
+	var setBuffer bytes.Buffer
+	setBuffer.WriteString(set)
+
+	tmpDir := os.TempDir() + "/"
+
+	err = os.Remove(tmpDir + binlog)
+	if err != nil && !os.IsNotExist(err) {
+		return errors.Wrap(err, "remove temp file")
+	}
+
+	err = syscall.Mkfifo(tmpDir+binlog, 0666)
+	if err != nil {
+		return errors.Wrap(err, "make named pipe file error")
+	}
+
+	file, err := os.OpenFile(tmpDir+binlog, syscall.O_NONBLOCK, os.ModeNamedPipe)
+	if err != nil {
+		return errors.Wrap(err, "open named pipe file error")
+	}
+	defer func() {
+		err = file.Close()
+		if err != nil {
+			err = errors.Wrapf(err, "close tmp file for %s", binlog)
+			return
+		}
```

Comment:
If this func (manageBinlog) happened to be exiting with error and you'll rewrite that error here with `nil` or some error from  file.Close() or os.Remove()